pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -326,16 +326,63 @@ func (c *Cluster) postReconcileUpdateOSDProperties(desiredOSDs map[int]*OSDInfo)
 	}
 	logger.Debugf("post processing osd properties with %d actual osds from ceph osd df and %d existing osds found during reconcile", len(osdUsage.OSDNodes), len(desiredOSDs))
 	for _, actualOSD := range osdUsage.OSDNodes {
-		if desiredOSD, ok := desiredOSDs[actualOSD.ID]; ok {
-			if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
-				// Log the error and allow other updates to continue
-				logger.Error(err)
-			}
+		if err := c.resizeOsdCrushWeight(actualOSD); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Errorf("failed to resize osd crush weight on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
+		}
+		desiredOSD, ok := desiredOSDs[actualOSD.ID]
+		if !ok {
+			continue
+		}
+		if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Errorf("failed to update device class on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
 		}
 	}
 	return nil
 }
 
+func convertKibibytesToTebibytes(kib string) (float64, error) {
+	kibFloat, err := strconv.ParseFloat(kib, 64)
+	if err != nil {
+		return float64(0), errors.Wrap(err, "failed to convert string to float")
+	}
+	return kibFloat / float64(1024*1024*1024), nil
+}
+
+func (c *Cluster) resizeOsdCrushWeight(actualOSD cephclient.OSDNodeUsage) error {
+	if !c.spec.Storage.AllowOsdCrushWeightUpdate {
+		return nil
+	}
+	crushWeight, err := strconv.ParseFloat(actualOSD.CrushWeight.String(), 64)
+	if err != nil {
+		return errors.Wrapf(err, "failed converting string to float for osd.%d crush weight %q", actualOSD.ID, actualOSD.CrushWeight.String())
+	}
+	// actualOSD.KB is in KiB units
+	desiredOSDTiB, err := convertKibibytesToTebibytes(actualOSD.KB.String())
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert KiB to TiB for osd.%d crush weight %q", actualOSD.ID, actualOSD.KB.String())
+	}
+	// do not reweight if the desired crush weight is 0 or less than the current crush weight
+	if desiredOSDTiB == float64(0) || desiredOSDTiB < crushWeight {
+		err := errors.Errorf("osd.%d desired crush weight %f less than actual crush weight %f or zero", actualOSD.ID, desiredOSDTiB, crushWeight)
+		return err
+	}
+	if desiredOSDTiB == crushWeight {
+		return nil
+	}
```

Comment:
As Parth brought up, we will likely need to check if the current crush weight is "close enough" to the calculated weight instead of doing an exact comparison.

I think we should get Ceph RADOS team input to help decide what "close enough" means. Rounding could be fine, but rounding will have different percentage effects for small disks compared to large disks. 0.005 (assuming rounding to decimal place) is a bigger percentage of a 500GB disk than a 30TB disk. Rounding to significant figures could have different impacts also.

Based on the numbers in the test Parth shared, there is ~0.07% difference between what we calculate vs what Ceph calculates when the disk is 40GB

```	{"id":0,"device_class":"hdd","name":"osd.0","type":"osd","type_id":0,"crush_weight":0.039093017578125,"depth":2,"pool_weights":{},"reweight":1,"kb":41943040,"kb_used":27640,"kb_used_data":432,"kb_used_omap":1,"kb_used_meta":27198,"kb_avail":41915400,"utilization":0.065898895263671875,"var":0.99448308946989694,"pgs":9,"status":"up"},
```

The difference is more like 3.02% and 2.4% from the user here, where large 1.8TB and 3.2TB disks are involved: https://forum.proxmox.com/threads/howto-define-osd-weight-in-crush-map.57469/

In the Proxmox cluster, rounding would still result in us adjusting the weight, which seems like we should try to avoid for the weights that Ceph sets by itself.

My instinct would be to suggest that we write a bit of code that determines the percent difference between the two values, and then we compare that to some threshold to determine if we should actually reweight or not. Even 2% difference seems like it could be a lot in a Ceph cluster, but maybe not. Maybe 5% would be better. I wouldn't go higher than 7% because if someone increases from using 28TB disks to 30TB disks -- which seems like the smallest reasonable increase for someone to make IMO -- that is a 7.14% increase.

We could consider having the threshold be user specifiable in the CephCluster. Instead of a boolean, we could take a float percentage value beyond which the weight will be adjusted automatically.

@travisn @parth-gr wdyt?

Reply:
Can we use an epsilon value that is just a raw number, instead of a percentage? For example, if we want to skip the update if the value is less than a 1GB diff, it would be something like this:
```
if Math.Abs(desiredOSDTiB - crushWeight) < 0.001 {
  // skip the update since it is unchanged or less than 1GB diff
  return nil
}
```

```

Reply:
> > Can we use an epsilon value that is just a raw number, instead of a percentage? For example, if we want to skip the update if the value is less than a 1GB diff, it would be something like this
> 
> I'm not sure we can, at least not with that low of a number. In the proxmox link, the person has a 1.8TB disk, and Ceph gives that a value of 1.627229. But the calculated value would be 1.676380634307861. That's 0.0492 off, about 50GB.
> 
> We could still use a flat number, but I would think we would want it to be more like ~250GB. Maybe 100GB at minimum. That would allow a reasonable 500GB->750GB disk upgrade (or 500GB->600GB). But I still think that some kind of user tunability might eventually be important to someone.
> 
> Tunability might also be practically necessary to do e2e testing of the feature, since test environments might only want to change 1GB or a few GB of space, instead of 100+.

Ah, I see now why the percent is important. There must be some difference in how ceph is calculating it if we are off more for bigger disks. 

@rzarzynski How does Ceph calculate the initial `weight` of the OSD? Is it not just the size of the disk divided by TiB? Or is there some other factor involved related to the CRUSH hierarchy?

If we can calculate it more consistently with Ceph, a 1% threshold would be good IMO, which would look like this:
```go 
if math.Abs((desiredOSDTiB / crushWeight) - 1) < 0.01 {
  // skip the update since it is unchanged or less than 1% diff
  return nil
}
```

Reply:
Hi @travisn! I think this snippet is what you're looking for:

```cpp
int OSD::update_crush_location()
{
  if (!cct->_conf->osd_crush_update_on_start) {
    dout(10) << __func__ << " osd_crush_update_on_start = false" << dendl;
    return 0;
  }

  char weight[32];
  if (cct->_conf->osd_crush_initial_weight >= 0) {
    snprintf(weight, sizeof(weight), "%.4lf", cct->_conf->osd_crush_initial_weight);
  } else {
    struct store_statfs_t st;
    osd_alert_list_t alerts;
    int r = store->statfs(&st, &alerts);
    if (r < 0) {
      derr << "statfs: " << cpp_strerror(r) << dendl;
      return r;
    }
    snprintf(weight, sizeof(weight), "%.4lf",
             std::max(.00001,
                      double(st.total) /
                      double(1ull << 40 /* TB */)));
  }

  std::multimap<string,string> loc = cct->crush_location.get_location();
  dout(10) << __func__ << " crush location is " << loc << dendl;

  string cmd =
    string("{\"prefix\": \"osd crush create-or-move\", ") +
    string("\"id\": ") + stringify(whoami) + string(", ") +
    string("\"weight\":") + weight + string(", ") +
    string("\"args\": [");
  for (multimap<string,string>::iterator p = loc.begin(); p != loc.end(); ++p) {
    if (p != loc.begin())
      cmd += ", ";
    cmd += "\"" + p->first + "=" + p->second + "\"";
  }
  cmd += "]}";

  return mon_cmd_maybe_osd_create(cmd);
}
```

Reply:
We are doing the same think 
Not sure why there is a difference 

```
double(st.total) /
                      double(1ull << 40 /* TB */)
```

Maybe we should just look for a 4 digit decimal precision

Reply:
@rzarzynski
Having a disk of 4Gib, ceph converts to `41943040`Kib.

But when it calcualtes the crush weight by converting `41943040` Kib to Tib it provides `0.039093017578125`

But in actually calculation `41943040` Kib should be `0.0390625`

So can we conclude ceph provides 4th decimal precision??

Reply:
Similarly if there is 10Gib disk, it converts it to `0.009796`Tib

But the actually calcluation would be 10Gib = `0.00976563` Tib or `0.009766` Tib

Reply:
WHat I see after 4th decimal place it adds `9` digit.

Reply:
I did some tests in minikube and am only seeing a diff of 0.01% of our calculation compared to Ceph's:
```
100GB —> “crush_weight":0.09768676757812
  Short version: 0.09769
  Actual version: 0.097656
  Diff: 0.01%

4TB —> "crush_weight":3.90618896484375
  Short version: 3.90619
  Actual Version: 3.90625
  Diff: 0.01%

40TB —> "crush_weight":40
  Short version: 40
  Actual calculated: 40
  Diff: 0
```
I don't actually see a problem here. We can use a small threshold such as 1% diff to determine if the disk is resized and we need to adjust the weight. I wonder if we were getting confused on a previous calculation around the kb, which was since resolved.

If we can use an accurate threshold like 1%, then we can keep the bool setting and no need to make the percentage configurable.

Reply:
It sounds good to me to go ahead with a small percent diff like this after that testing. 1% is still 10x more than "actual", which is quite a bit of padding for error. 👍